packages/dds/tree/src/core/rebase/utils.ts

```diff
@@ -0,0 +1,139 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+
+/**
+ * Find an ancestor of some descendant.
+ * @param descendant - a descendant. If an empty `path` array is included, it will be populated
+ * with the chain of ancestry for `descendant` from most distant to closest (including `descendant`,
+ * but not including the ancestor found by `predicate`).
+ * @param predicate - a function which will be evaluated on every ancestor of `descendant` until it returns true.
+ * @returns the closest ancestor of `descendant` that satisfies `predicate`, or `undefined` if no such ancestor exists.
+ * @example
+ * ```ts
+ * interface Parented {
+ *   id: string;
+ *   parent?: Parented;
+ * }
+ * const g = { id: "g" }; // Grandparent
+ * const p = { parent: g, id: "p" }; // Parent
+ * const c = { parent: p, id: "c" }; // Child
+ * const path: Parented[] = [];
+ * const ancestor = findAncestor<Parented>([c, path], (n) => n.id === "g");
+ * // ancestor === g
+ * // path === [p, c]
+ * ```
+ */
+export function findAncestor<T extends { parent?: T }>(
+	descendant: T | [descendant: T, path?: T[]] | undefined,
+	predicate: (t: T) => boolean,
+): T | undefined {
+	let d: T | undefined;
+	let path: T[] | undefined;
+	if (Array.isArray(descendant)) {
+		[d, path] = descendant;
+	} else {
+		d = descendant;
+	}
+	for (let cur = d; cur !== undefined; cur = cur.parent) {
+		if (predicate(cur)) {
+			return cur;
+		}
+		path?.unshift(cur);
+	}
+
+	if (path !== undefined) {
+		path.length = 0;
+	}
+	return undefined;
+}
+
+/**
+ * Find a common ancestor between two descendants that are linked by parent pointers.
+ * @param descendantA - a descendant. If an empty `path` array is included, it will be populated
+ * with the chain of ancestry for `descendantA` from most distant to closest (including `descendantA`,
+ * but not including the ancestor shared with `descendantB`).
```

Comment:
Consider mentioning that `descendantA` will not be included if it happens to be the ancestor of (or equal to) `descendantB`.
Same for the converse.

Reply:
Okay I tried to say it a different way. If somebody is still not completely confident about what it means, hopefully the example will be enough to solidify it.